Customizable scripts through .env files.

danielaRiesgo opened the pull request:
Adding READMEFIRST.md with how to use scripts with all explanation.

Making scripts take information form env files for each environment (and one general .env).

It was tested in GreenCo and it's working.

pablog commented on Fastfile.private:
```diff
@@ -58,36 +61,42 @@ reflected in `Info.plist` during building."
 
-    build_configuration_key = options[:build_configuration]
-    build_configuration = get_build_configuration(build_configuration: build_configuration_key)
+    environment = options[:environment]
```
I'd keep the `_key` to distinguish between the key name and the actual environment.

pablog replied:
I'll leave a comment just here, but it's to apply in every place this was changed.

danielaRiesgo replied:
I don't get it. Now build_configuration_key is environment and build_configuration is buikld_configuration.
`:environment` is not a variable, just a symbol, like "environment".

pablog replied:
Yes, sure, ignore my comment.

pablog commented on Fastfile.private:
```diff
@@ -196,20 +208,22 @@ reflected in `Info.plist` during building."
       p12_password: p12_password,
-      output_path: output_path
+      output_path: output_path,
+      username: credentials[:account],
```
I'd make the names match the expected ones: "username" -> "username", "team_id" -> "team_id".

danielaRiesgo replied:
the thing is that sometime itc_team_id is team_id because the other team_id isn't necessary for that lane. Or sometimes is username and sometime I don't remember what else.
So I tried to give them an understandable name despite the context for what it will be used.
Like the itc_team is the team from itunes connect, and account si the account by which to login to apple.

pablog commented on Fastfile.private:
```diff
@@ -220,6 +234,7 @@ reflected in `Info.plist` during building."
       # http://krausefx.com/blog/download-dsym-symbolication-files-from-itunes-connect-for-bitcode-ios-apps
       include_bitcode: false
     )
+    ### Check if external_team_id needed
```
What do you mean?

pablog commented on Fastfile.private:
```diff
@@ -253,21 +272,31 @@ reflected in `Info.plist` during building."
+      clone_branch_directly: true,
+      username: credentials[:account],
+      team_id: credentials[:team],
+      git_url: certificates_info[:url],
```
Same here, make explicit `url` means `git_url`

danielaRiesgo commented on Fastfile.private:
```diff
@@ -8,20 +8,23 @@ platform :ios do
-    expected_bundle_identifiers = Hash[build_configurations.values.uniq.map { |each| [each, []] }]
+    build_configurations = Actions::GetBuildConfigurationAction.get_all_build_configurations
+
+    expected_bundle_identifiers = Hash[build_configurations.values.uniq.map { |each| [ each, [nil,[]] ] }]
```
Yes ! I just don't know how to do a struct. I'll find out and change it.

pablog replied:
If you just add a comment explaining what it's doing it's enough for me.

pablog commented on Fastfile.private:
```diff
@@ -222,21 +222,29 @@ reflected in `Info.plist` during building."
-    )
-    ### Check if external_team_id needed
+    }
+    if project_file.include? ".xcodeproj" do
```
There's always a ".xcodeproj" file. Better ask for `if project_file.include? ".xcworkspace" do` and do the opposite.

pablog replied:
Anyway, I'm not sure what you mean by doing this. Why is this necessary?

danielaRiesgo replied:
Because in `gym` call, you must specify the project or the workspace with a different parameter (I guess that if you don't it infers it, but maybe it infers an incorrect one. In any case is better to make sure it uses the one set up in the configuration).
The action `project_file` gets the project file name according to configurations. So if the file name has a .xcodeproj, it is an xcodeproj, if not it's a workspace (or the other way around, it's the same)

pablog replied:
The point is that the `.xcodeproj` is always present, even if there's also a `.xcworkspace`. This is why I'd do the opposite.

pablog commented on Fastfile.private:
```diff
@@ -239,10 +239,10 @@ reflected in `Info.plist` during building."
     else
-      params[:scheme] = project_file
+      params[:project] = project_file
     end
```
I'd add a new `else` to handle if there's no `. xcworkspace` and no `. xcodeproj`, which means the project is misconfigured. In this case you can do `UI.error('No Xcode project file found')` or something like this.

pablog commented:
@danielaRiesgo can we resume this PR? is there anything you need to finish it?

danielaRiesgo commented:
@pablog Yes, I need to add the checks and settings and then it will be ready, but I didn't have time. It's my next move.

danielaRiesgo commented:
@pablog @jcaracciolo Summary
- Added .env files to set information on the environments.
- Added configuration of many parameters like scheme that wasn't treated before.
- I added confirmation of data, and then it is set by the scripts into the xcproject.

It was tested in GreenCommuter and was working.

My only concern, is that if you don't have the 4 environments (like in RN), you will be changing the xcodeproj file everytime you deploy, and so you would have to commit those changes and probably you'll have merging problems all the time I guess. I'm thinking of doing a default lane, that for each build_configuration it chooses the first environment that appears and sets all that info (which will be overwritten everytime a lane starts, but then set back to that default everytime a lane finishes).

danielaRiesgo commented:
What do you mean by `will keep working after this`?
For example, after running `release_qa` lane, in RN the "Release" build configuration will end up with all qa information (bundle id, development team, provisioning profile, etc), which "will work" for qa.
In iOS, after running `release_qa`, the QA build configuration will end up with all QA information, so that won't change anything.

pablog commented:
@danielaRiesgo 

```
@pablog What do you mean by will keep working after this?
```

I mean we'll have the same features we currently have or more.

```
For example, after running release_qa lane, in RN the "Release" build configuration will end up with all qa information (bundle id, development team, provisioning profile, etc), which "will work" for qa.
In iOS, after running release_qa, the QA build configuration will end up with all QA information, so that won't change anything.
```

We have a workaround for that in RN. Since we have only one build configuration for more than one lane, we keep one as default and replace values when when use the other one.

For example, we default `Release` for QA, and replace values for external appstore.

Check if you want to build a helper function to make this easier in RN (this is copied from `Carvi`):

```
  lane :release_external_appstore do |options|
    `mv ../../.env ../../.env.bkp`
    `mv ../../.env.production ../../.env`
    update_project_property(
      build_configuration: get_build_configuration(build_configuration: :production),
      build_setting: 'PRODUCT_BUNDLE_IDENTIFIER',
      build_setting_value: 'com.Carvi.Carvi'
    )
    update_project_property(
      build_configuration: get_build_configuration(build_configuration: :production),
      build_setting: 'PROVISIONING_PROFILE_SPECIFIER',
      build_setting_value: 'match AppStore com.Carvi.Carvi'
    )
    update_project_property(
      build_configuration: get_build_configuration(build_configuration: :production),
      build_setting: 'DEVELOPMENT_TEAM',
      build_setting_value: 'Z4R3CWY2Z3'
    )
    release build_configuration: :production, bump_type: options[:bump_type]
    update_project_property(
      build_configuration: get_build_configuration(build_configuration: :production),
      build_setting: 'PRODUCT_BUNDLE_IDENTIFIER',
      build_setting_value: 'com.wolox.Carvi.qa'
    )
    update_project_property(
      build_configuration: get_build_configuration(build_configuration: :production),
      build_setting: 'PROVISIONING_PROFILE_SPECIFIER',
      build_setting_value: 'match AppStore com.wolox.Carvi.qa'
    )
    update_project_property(
      build_configuration: get_build_configuration(build_configuration: :production),
      build_setting: 'DEVELOPMENT_TEAM',
      build_setting_value: '6LGQ9AZEQX'
    )
    `mv ../../.env ../../.env.production`
    `mv ../../.env.bkp ../../.env`
  end
```

danielaRiesgo commented:
@pablog 
Yeah, we will have the same info.
And yes, I was thinking of something like that, but in how to do it generic.

pablog commented:
I'd make an action called `replace_build_configuration_values` (or similar) which receives:

- bundle_identifier
- provisioning_profile_specifier
- development_team
- env_file_path

All of them optionals, and in case they're not present just don't replace.

danielaRiesgo commented:
Tested in Papa Johns !
